[MRG] DOC: Update contributing guide and windows install

ntolley opened the pull request:
More verbose description of the pull request process for new contributors.

jasmainak commented on doc/contributing.rst:
```diff
@@ -14,6 +14,115 @@ If your contributions will make use of parallel backends for using more than
 one core, please see the additional installation steps in our
 :doc:`parallel backend guide <parallel>`.
 
+Overview of contribution process
```
I think if you use some of the wording from this section to make the next section a little more verbose, one could get rid of the section altogether.

rythorpe commented on doc/contributing.rst:
```diff
@@ -14,6 +14,115 @@ If your contributions will make use of parallel backends for using more than
+Instructions for how to configure git can be found on 
+the [git book](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration).
+
+Making a pull request
```
```suggestion
Making your first pull request
```

jasmainak commented:
Not all the blocks of commands have rendered correctly. See: https://output.circle-artifacts.com/output/job/ec1d9318-79d4-4e21-8f6e-2916a48e8a06/artifacts/0/html/contributing.html

![image](https://user-images.githubusercontent.com/15852194/184244748-d5f94bd3-08c0-474d-a489-6dbb0ce66aad.png)

You might be missing a colon or two

ntolley commented:
@jasmainak @rythorpe I updated the windows install instructions. The rendering issues on the contributing guide should be fixed now too

Besides a linkcheck error I need to resolve, I'm pretty set with this PR. Let me know if there's any changes you'd like to see!

jasmainak commented on README.rst:
```diff
@@ -76,18 +76,29 @@ than one CPU core, refer to our `parallel backend guide`_.
+install Neuron using the traditional package installer available here
 `https://neuron.yale.edu/neuron/ <https://neuron.yale.edu/neuron/>`_.
 
+manually install ``hnn-core`` by cloning the latest stable release from github::
+   $ git clone https://github.com/jonescompneurolab/hnn-core --depth 1
+   $ cd hnn-core
+   $ git fetch --tags
+   $ latestTag=$(git describe --tags `git rev-list --tags --max-count=1`)
+   $ git checkout $latestTag
+
+and finally pip install the local copy, as well as build the Neuron .mod files::
+   $ pip install -e .
+   $ python setup.py build_mod
```
I'm tempted to just nuke this from the instructions ... until we figure out a simpler solution.

ntolley replied:
Alternatively, it acts as a pretty good deterrent against trying the Windows install ;)

jasmainak replied:
we don't want to lose 25% of our customers:

![image](https://user-images.githubusercontent.com/15852194/186768527-a99da3c3-f178-4792-98c5-edadfa4c80b6.png)

from the pre-workshop survey

jasmainak commented on README.rst:
```diff
@@ -76,18 +76,29 @@ than one CPU core, refer to our `parallel backend guide`_.
-The pip installer for Neuron does not yet work for Windows. In this case, it is better to
-install ``hnn_core`` without the dependencies::
+The pip installer for Neuron does not yet work for Windows. In this case we recommend
+installing WSL on your local machine, and install ``hnn-core`` using the same steps as
```
provide a link to WSL installer?

rythorpe commented:
I'll try to review in detail this evening!

jasmainak commented:
merge button is yours @rythorpe !

rythorpe commented on README.rst:
```diff
@@ -76,18 +76,31 @@ than one CPU core, refer to our `parallel backend guide`_.
    $ pip install scipy numpy matplotlib
 
-and install Neuron using the traditional package installer available here
+install Neuron using the traditional package installer available here
```
I thought the pip installation for Neuron works, but it's just that the mod files need to be build by calling them manually?

ntolley replied:
Oh wow I think you're right, that is actually new. Do you have a windows machine you can verify this with?

rythorpe replied:
I was thinking that it was possible all along, but that's probably my ignorance talking. I could test it out tonight, but it might be prudent to merge this PR before then.

jasmainak replied:
I don't think Neuron can be installed on Windows with pip ...

rythorpe commented on doc/contributing.rst:
```diff
@@ -14,11 +14,97 @@ If your contributions will make use of parallel backends for using more than
+   Note: see the `numpy contributing guide <https://numpy.org/doc/stable/dev/development_workflow.html#writing-the-commit-message>`_
+   for tips on informative commit messages.
+
+8. Now push the feature branch to our fork::
```
```suggestion
8. Now push the feature branch to your fork::
```

rythorpe commented on doc/contributing.rst:
```diff
@@ -14,11 +14,97 @@ If your contributions will make use of parallel backends for using more than
+    $ git push origin cool_feature
+
+9. Go to https://github.com/jonescompneurolab/hnn-core/compare to open the pull request by selecting your feature branch.
+   You should also see this link on the terminal when you make the push. 
```
Which link is supposed to show up on the terminal when you push? I'm not sure I follow what the point is here...

ntolley replied:
I think it's just to draw attention to the fact that an automatic message gets generated about opening a PR? It's not exactly this link (the link check will fail on the PR otherwise) but they serve the same purpose

rythorpe commented on doc/contributing.rst:
```diff
@@ -14,11 +14,97 @@ If your contributions will make use of parallel backends for using more than
-repository with an editable pip install::
+To run the tests using ``pytest``, you need to have the git cloned ``hnn-core`` repository with an editable pip install::
 
     $ git clone https://github.com/jonescompneurolab/hnn-core --depth 1
```
I've always wondered this: is the `--depth 1` argument necessary?

ntolley replied:
This just prevents downloading the entire commit history, it's still accessible if they do a git fetch

rythorpe replied:
Okay, that's what I thought. My super nitpicky suggestion would be to remove this if it's not necessary because simplicity is better, especially when trying to onboard a new contributor who is still trying to wrap their head around the git workflow.

jasmainak commented on README.rst:
```diff
@@ -76,18 +76,31 @@ than one CPU core, refer to our `parallel backend guide`_.
-The pip installer for Neuron does not yet work for Windows. In this case, it is better to
-install ``hnn_core`` without the dependencies::
+The pip installer for Neuron does not yet work for Windows. In this case we recommend
+installing `WSL <https://docs.microsoft.com/en-us/windows/wsl/install>`_ on your
```
It seems that it's not obvious Anaconda needs to be installed also on WSL. Clarifying may help!

jasmainak commented:
So just as an update, I met with two users on zoom today to debug their installation issues ... the Mac users got it installed in 15 minutes, for the Windows users I couldn't figure out the issue after 1.5 hours of debugging. WSL didn't help. All sorts of issues, related to display and also having the right ipython version. I ended up recommending Linux

jasmainak commented:
@ntolley let me know when this is ready and you've addressed @rythorpe 's comments. Let's try to merge before the workshop

ntolley commented:
@jasmainak I think it's best to put this back to WIP. I'd say the contributing guide is fairly low priority and not necessary to merge before the workshop. Ryan and I will troubleshoot the windows install on his machine when I return so we can have a solid set of install instructions

jasmainak commented on README.rst:
```diff
@@ -76,18 +76,31 @@ than one CPU core, refer to our `parallel backend guide`_.
+   $ latestTag=$(git describe --tags `git rev-list --tags --max-count=1`)
+   $ git checkout $latestTag
+
+and finally pip install the local copy, as well as build the Neuron .mod files::
+   
+   $ pip install -e .
+   $ python setup.py build_mod
```
I don't think we should be encouraging regular users to do editable installs. You'll start to see users who messed up their local directory with git and then you have to teach them git :) The right fix is to update the MANIFEST.in and package_data directory correctly. The questions to answer are:

1. If you install Neuron on Windows, can it be installed within an Anaconda environment? See [here](https://web.archive.org/web/20201126015939/https://www.neuron.yale.edu/ftp/neuron/nrn_mswin_install.pdf)
2. If the mod files are built on Windows, what's the best way to copy the dll file to the Anaconda directory.

Step 2. is what doesn't get run in an editable install. We should fix it instead of suggesting users to do an editable install.

jasmainak commented:
See my comments with regards to editable install. We need to do a "proper fix". But I'm okay to merge just the contributing guide and keeping a WIP PR open for the rest.

ntolley commented:
@jasmainak @rythorpe I think I've addressed all the comments! Let me know if everything looks good to you once the docs merged. I just decided to say we don't support windows install for now. This might change by the time of the release but given our current experiences I think it's accurate to say for now.

jasmainak commented:
I let @rythorpe merge if happy

rythorpe commented:
Thanks @ntolley!
